### PR TITLE
Grid Lines

### PR DESCRIPTION
**Pull request by bfox6**
**Work Completed**
- Added a CustomGridLines component that renders an array of Leaflet rectangles representing grid lines.
- Added a function, ZoomLevelToGeoHashLevel, that will return a geohash precision when given a zoom level.
- Added shape2geohash library

**Required Steps**
- Reinstall libraries

### CODE REVIEW

**Review by bobular — COMMENTED** [review]
Looks great all round, Brett.

My only suggestion would be to add a prop to MapVEuMap to make the grid switchable on/off.  I don't know what our prop naming strategy is but `showGrid` = true/false works for me.

Also, please could you investigate the buggy merged rectangles at widest zoom level?  Perhaps use filled rectangles of random colors to check they are really there?

No need to implement ZoomLevelToGeoHashLevel function (though should the function name start with a lowercase letter?) in src/map/AnimatedMarkers.stories.tsx - treat all story code as disposable ;-)

**Review by dmfalke — APPROVED** [review]
Overall, this looks good!

I did notice some typescript errors. I'm not sure where they originated (here or an upstream branch?), but they should get fixed.

**Comment by dmfalke** [comment]
Here is the output from `tsc`:

```
❯ yarn tsc --pretty
yarn run v1.22.5
$ /home/dfalke/Work/mapveu/node_modules/.bin/tsc --pretty
src/map/AnimatedMarkers.stories.tsx:9:17 - error TS7016: Could not find a declaration file for module 'md5'. '/home/dfalke/Work/mapveu/node_modules/md5/md5.js' implicitly has an 'any' type.
  Try `npm install @types/md5` if it exists or add a new declaration (.d.ts) file containing `declare module 'md5';`

9 import md5 from 'md5';
                  ~~~~~

src/map/AnimatedMarkers.stories.tsx:152:6 - error TS2741: Property 'animation' is missing in type '{ viewport: { center: number[]; zoom: number; }; height: string; width: string; onViewportChanged: (bvp: BoundsViewport, duration: number) => void; markers: ReactElement<MarkerProps, string | ... 1 more ... | (new (props: any) => Component<...>)>[]; }' but required in type 'MapVEuMapProps'.

152     <MapVEuMap
         ~~~~~~~~~

  src/map/Types.ts:44:3
    44   animation: {
         ~~~~~~~~~
    'animation' is declared here.

src/map/AnimatedMarkers.stories.tsx:204:23 - error TS2554: Expected 3-4 arguments, but got 2.

204     setMarkerElements(getMarkerElements(bvp, 100000));
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  src/map/AnimatedMarkers.stories.tsx:39:88
    39 const getMarkerElements = ({ bounds, zoomLevel }: BoundsViewport, numMarkers : number, duration: number, scrambleKeys: boolean = false) => {
                                                                                              ~~~~~~~~~~~~~~~~
    An argument for 'duration' was not provided.

src/map/AnimatedMarkers.stories.tsx:213:5 - error TS2322: Type '{ viewport: { center: number[]; zoom: number; }; height: string; width: string; onViewportChanged: (bvp: BoundsViewport) => void; markers: ReactElement<MarkerProps, string | ((props: any) => ReactElement<...> | null) | (new (props: any) => Component<...>)>[]; setMarkerElements: Dispatch<...>; }' is not assignable to type 'IntrinsicAttributes & MapVEuMapProps'.
  Property 'setMarkerElements' does not exist on type 'IntrinsicAttributes & MapVEuMapProps'.

213     setMarkerElements={setMarkerElements}
        ~~~~~~~~~~~~~~~~~

src/map/MapVEuMap.stories.tsx:39:6 - error TS2741: Property 'animation' is missing in type '{ viewport: { center: number[]; zoom: number; }; height: string; width: string; onViewportChanged: (bvp: BoundsViewport) => void; markers: ReactElement<MarkerProps, string | ((props: any) => ReactElement<...> | null) | (new (props: any) => Component<...>)>[]; }' but required in type 'MapVEuMapProps'.

39     <MapVEuMap
        ~~~~~~~~~

  src/map/Types.ts:44:3
    44   animation: {
         ~~~~~~~~~
    'animation' is declared here.


Found 5 errors.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Review by bobular — COMMENTED** [review]
Because we're using only the north east corner, there are grid lines "missing" on the south and western edges (when zoomed fully out).

Is the east longitude of one geohash the same as the west longitude of its neighbour?

Maybe we can just throw both boundaries (top and bottom or left and right) into the uniqueness check?

File: src/map/CustomGridLayer.tsx
Comment (dmfalke) [inline src/map/CustomGridLayer.tsx:61]: This logic doesn't need to be in a `useEffect` hook. Since the code is all synchronous, you can just create the `latLines` and `lonLines` variables directly (no need for `useState`).

**Review by dmfalke — COMMENTED** [review]
I noticed a couple of unnecessary uses of hooks, and a place where using `Array.prototype.map` can be changed to using `Array.prototype.forEach` (see inline comments).

Otherwise, this looks good!

**Comment by dmfalke** [comment]
This all looks great to me. I'll let @bobular decide when if it's ready to me merged.

**Review by bobular — APPROVED** [review]
Looks great now.

**Comment by bobular** [comment]
Seems to work fine after my merge conflict resolution :-)

**Pull request merged**
